doc/kb_sdk_impl_methods.md

```diff
@@ -0,0 +1,201 @@
+
+```python
+        # load the method provenance from the context object
+        provenance = [{}]
+        if 'provenance' in ctx:
+            provenance = ctx['provenance']
+        # add additional info to provenance here, in this case the input data object reference
+        provenance[0]['input_ws_objects']=[]
+        provenance[0]['input_ws_objects'].append(params['workspace_name']+'/'+params['read_library_name'])
+```
+
+[\[Back to Edit Impl\]](#impl)
+
+##### <A NAME="impl-report"></A>5.5. Building Output Report
+
+```python
+        # create a Report
+        report = ''
+        report += 'ContigSet saved to: '+params['workspace_name']+'/'+params['output_contigset_name']+'\n'
+        report += 'Assembled into '+str(len(contigset_data['contigs'])) + ' contigs.\n'
+        report += 'Avg Length: '+str(sum(lengths)/float(len(lengths))) + ' bp.\n'
+
+        # compute a simple contig length distribution
+        bins = 10
+        counts, edges = np.histogram(lengths, bins)
+        report += 'Contig Length Distribution (# of contigs -- min to max basepairs):\n'
+        for c in range(bins):
+            report += '   '+str(counts[c]) + '\t--\t' + str(edges[c]) + ' to ' + str(edges[c+1]) + ' bp\n'
+
+        reportObj = {
+            'objects_created':[{'ref':params['workspace_name']+'/'+params['output_contigset_name'], 'description':'Assembled contigs'}],
+            'text_message':report
+        }
+
+        reportName = 'megahit_report_'+str(uuid.uuid4())   # uuid4() instead of getnode() to ensure uniqueness
+        report_obj_info = ws.save_objects({
+                'id':info[6],
+                'objects':[
+                    {
+                        'type':'KBaseReport.Report',
+                        'data':reportObj,
+                        'name':reportName,
+                        'meta':{},
+                        'hidden':1,
+                        'provenance':provenance
+                    }
+                ]
+            })[0]
+
+        output = { 'report_name': reportName, 'report_ref': str(report_obj_info[6]) + '/' + str(report_obj_info[0]) + '/' + str(report_obj_info[4]) }
+```
+
+[\[Back to Edit Impl\]](#impl)
+
+##### <A NAME="impl-shell-tool"></A>5.6. Invoking Shell Tool
+
+```python
+        command_line_tool_params_str = " ".join(command_line_tool_params)
+        command_line_tool_cmd_str = " ".join(command_line_tool_path, command_line_tool_params_str)
+        
+        # run <command_line_tool>, capture output as it happens
+        self.log(console, 'running <command_line_tool>:')
+        self.log(console, '    '+command_line_tool_cmd_str))
+        p = subprocess.Popen(command_line_tool_cmd_str,
+                    cwd = self.scratch,
+                    stdout = subprocess.PIPE, 
```

Comment:
Why bother with piping things around? Just use call and things will happen as they should.